Update: added empty error array check for false negative

anikethsaha opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).
- [x] The team has reached consensus on the changes proposed in this pull request. If not, I understand that the evaluation process will begin with this pull request and won't be merged until the team has reached consensus.

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[X] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)
Currently in test files for rules and internal rules, if a valid test case in added to invalid test case and have `errors` set to an empty array, it simply passes the test which should be an error as it is false negative. 
Added a check for empty-ness of the array and it will throw  error if the `errors` is empty array

#### Is there anything you'd like reviewers to focus on?
~None~
not sure whether it is of type `Fix` or `Chore` cause I think it is addressing false negative of test cases. let me know 👍

aladdin-add commented on lib/rule-tester/rule-tester.js:
```diff
@@ -644,6 +644,10 @@ class RuleTester {
                 `Did not specify errors for an invalid test of ${ruleName}`);
 
+            if (Array.isArray(item.errors) && item.errors.length === 0) {
+                assert.fail("errors array should have atleast one element otherwise its a valid case");
```
`atleast` => `at least`

mdjermanovic commented:
Interesting, I can confirm that test cases such as this one will indeed pass in the actual version:

```js
invalid: [
    {
        code: "some_valid_code_for_this_rule",
        errors: []
    }
]
```

I think we should fix this.

> not sure whether it is of type Fix or Chore cause I think it is addressing false negative of test cases.

RuleTester is public API, so this would be at least `Fix` or `Update`. I believe it doesn't have to be marked as a `Breaking` change, but would like more opinions.

mdjermanovic commented:
If we decide to fix RuleTester to fail on `errors: []`, maybe we should do the same for `errors: 0`.

anikethsaha commented:
> If we decide to fix RuleTester to fail on `errors: []`, maybe we should do the same for `errors: 0`.

true, I will add check for this as well. 

> RuleTester is public API, so this would be at least `Fix` or `Update`. I believe it doesn't have to be marked as a `Breaking` change, but would like more opinions.

I think fix might be better cause of false results .

kaicataldo commented:
Agreed that we should fix this. I would also support adding this as a semver-minor bug fix (or in our current major release cycle if this gets approved in time).

anikethsaha commented:
@mdjermanovic I added check for `errors: 0`. should it be better if the check was `< 1` idk any tests has negative numbers. 
@kaicataldo I changed the 1st commit message and PR title to `Update` type. let me know if anything else is required to change.

mdjermanovic commented:
> I added check for `errors: 0`. should it be better if the check was `< 1` , idk any tests has negative numbers.

This should be already caught when the tester compares the expected number with the number of actual errors, though it does make sense to check that explicitly. I don't have a strong opinion on this.

anikethsaha commented:
> This should be already caught when the tester compares the expected number with the number of actual errors, though it does make sense to check that explicitly.

I think we should keep that as the current one cause that will help more than just a message saying it can't be negative. 
so, i think its fine to just check for `0`

anikethsaha commented:
Anything else is needed for this  ?

cc @mdjermanovic

mdjermanovic commented:
> Anything else is needed for this ?
> 
> cc @mdjermanovic

The change looks good in general, let's wait to see if it will get accepted before reviewing details.

kaicataldo commented on lib/rule-tester/rule-tester.js:
```diff
@@ -644,13 +644,22 @@ class RuleTester {
                 `Did not specify errors for an invalid test of ${ruleName}`);
 
+            if (Array.isArray(item.errors) && item.errors.length === 0) {
+                assert.fail("errors array should have at least one element otherwise its a valid case");
```
```suggestion
                assert.fail("Invalid cases must have at least one error.");
```

kaicataldo commented on lib/rule-tester/rule-tester.js:
```diff
@@ -644,13 +644,22 @@ class RuleTester {
             if (typeof item.errors === "number") {
+
+                if (item.errors === 0) {
+                    assert.fail("errors of type `number` should not have value `0`");
```
This error message explains what but not why. Can this be the same error message as above, or is there maybe a better way to explain why this error message is being reported?

anikethsaha replied:
I changed as mentioned above.
 for `why` , is something like `Invalid cases must have value greater than 0 as having 0 can result in false positive` sounds better  ?

kaicataldo replied:
Could we just report the same message as above? `"Invalid cases must have at least one error"`?

anikethsaha replied:
umm,
atleast one error might not specify non-zero number, right ? cause here it should be > 0 if its of type `number` .

WDYT ?

kaicataldo replied:
I see. Can we explain why a false positive is possible? My concern here is that this error message doesn't give a lot of context around what the actual problem is, but instead describes the implementation detail.

anikethsaha replied:
What I think is, 

for error type `number` ,

if its `0`  (`items.errors = 0`), and for valid case `messages.length` is also `0` so we are doing `assert.strictEqual(message.lenght, items.errors)` which will be `true`

and for `Array<object>` type, 
it iterates over each, so if the case is `valid` , there won't be any iteration as  `item.errors.length` is `0` so there won't be any assertion (check) so it will be false positive. 
 
Now, we need to compress this to a short message for the explanation 😄.
cause we cant have these much long detail right ?
any suggestion for this would be awesome.

anikethsaha replied:
cc @kaicataldo thoughts !?

kaicataldo replied:
Sorry, will try to get to this next week!

anikethsaha replied:
cool 👍

kaicataldo commented:
Sorry for the delay!

kaicataldo commented:
@aladdin-add Can you confirm your concerns have been addressed?

mdjermanovic commented on lib/rule-tester/rule-tester.js:
```diff
@@ -644,13 +644,22 @@ class RuleTester {
             if (typeof item.errors === "number") {
+
+                if (item.errors === 0) {
+                    assert.fail("Invalid cases must have value greater than 0");
```
Maybe `"Invalid cases must have 'error' value greater than 0"`?
